File: apps/backend/internal/domain/song.go
package domain

import (
	"time"
)

type Song struct {
	ID           string    `gorm:"primaryKey;type:varchar(50)" json:"id"`
	Title        string    `gorm:"type:varchar(255)" json:"title"`
	ChannelTitle string    `gorm:"type:varchar(255)" json:"channelTitle"`
	ThumbnailURL string    `gorm:"type:text" json:"thumbnailUrl"`
	ViewCount    string    `gorm:"type:varchar(50)" json:"viewCount"`
	CountryCode  string    `gorm:"primaryKey;type:varchar(2)" json:"countryCode"`
	CategoryID   string    `gorm:"type:varchar(50)" json:"categoryId"`
	PublishedAt  time.Time `json:"publishedAt"`
	VideoType    string    `gorm:"type:varchar(50)" json:"videoType"`
}

type SongRepository interface {
	UpsertSongs(songs []Song) error
	GetTrends(countryCode string) ([]Song, error)
	GetNewReleases(countryCode string) ([]Song, error)
	GetMVs(countryCode string) ([]Song, error)
}

type YouTubeRepository interface {
	FetchTrending(countryCode string) ([]Song, error)
}

type SongUsecase interface {
	SyncTrendingMusic(countryCode string) error
	GetTrends(countryCode string) ([]Song, error)
	GetNewReleases(countryCode string) ([]Song, error)
	GetMVs(countryCode string) ([]Song, error)
}

File: apps/mobile/lib/models/song.dart
class Song {
  final String id;
  final String title;
  final String channelTitle;
  final String thumbnailUrl;
  final String viewCount;
  final String countryCode;
  final String categoryId;
  final String publishedAt;
  final String videoType;

  const Song({
    required this.id,
    required this.title,
    required this.channelTitle,
    required this.thumbnailUrl,
    required this.viewCount,
    required this.countryCode,
    required this.categoryId,
    required this.publishedAt,
    required this.videoType,
  });

  factory Song.fromJson(Map<String, dynamic> json) {
    return Song(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      channelTitle: json['channelTitle'] as String? ?? '',
      thumbnailUrl: json['thumbnailUrl'] as String? ?? '',
      viewCount: json['viewCount'] as String? ?? '0',
      countryCode: json['countryCode'] as String? ?? '',
      categoryId: json['categoryId'] as String? ?? '',
      publishedAt: json['publishedAt'] as String? ?? '',
      videoType: json['videoType'] as String? ?? 'General',
    );
  }
}
